README.md
# Tic Tac Toe

## Tech behind
This project was built using mongodb for the data storage and tornado web is the main framework. In order 
to take advantage of the async nature of the framework, umongo is being used on top of motor for mongo.

## REST API Documentation

We support three elements that are going to be stored: Users, Games, and GameMoves.

### Users
A user is a player of our tool. This is how a user looks in the database:
```json
{
    "_id" : ObjectId("5c9d5702e3872b02c94ecdb0"),
    "email" : "test@test.de",
    "username" : "usertest",
    "created_at" : ISODate("2019-03-29T00:21:29.287Z"),
    "victories" : 0
}
```

In order to manipulate users we have the next set of endpoints:

#### POST /api/users
```json

{
	"username": "test",
	"email": "test@test.de"
}
```
There are two main attributes, email and username, which are unique.


#### PUT /api/users/{user_id}
```json
{
	"email": "test@test.de"
}
```
The user_id is the mongo_id of the user when stored. Any of field can be modified by just sending 
the updated fields in the json body of the request

#### GET /api/users
Retrieves the full list of stored users.

#### GET /api/users/{user_id}
The user_id as explained before, is the object_id of the database entry


#### DELETE /api/users/{user_id}
It is a user hard delete. Soft delete will be for a later improvement.

### Games
A game is an object represented by a board, a set of players and a state. Each move on this board
will be stored in the GameMove collection, being able then to trace the user behavior on the game.
A game has a status, which might be: `created`, `in_progress`, `tie` or `finished`. Each of these states
will be updated accordingly to the game progress. 

A game entry will look like this in the database:
```json
{
    "_id" : ObjectId("5c9d59d3e3872b091268fa1b"),
    "players" : [ 
        ObjectId("5c9d59cfe3872b091268fa19"), 
        ObjectId("5c9d59d1e3872b091268fa1a")
    ],
    "status" : "created",
    "board" : [ 
        [ 
            "", 
            "", 
            ""
        ], 
        [ 
            "", 
            "", 
            ""
        ], 
        [ 
            "", 
            "", 
            ""
        ]
    ],
    "multiplayer" : true
}
```
The board is a 3 by 3 grid, that is represented by a matrix on the board field of the database entry
```
[["", "", ""],
 ["", "", ""],
 ["", "", ""]]
```
This should be the visual representation of the board. We have two available markers `X` or `O` and each cell 
of the grid will be replaced with those values when the game starts moving forward.

#### POST /api/games
```json
{
 "players": [
    "5c9d51d5e3872b71421ae42c",
    "5c9d50f5e3872b6fb001ae77"
  ]
}
```
This will create a new game board. The game will be multiplayer or single player depending on the amount of players sent
on the json body. A player is a mongo_id.

#### GET /api/games
Retrieve the full list of games stored


#### GET /api/games/{game_id}
game_id is the mongo_id of the database entry. When calling this endpoint it will return the current state
of the requested game.

### GameMoves
A game move is an object that describes the next requested move by the user. This move will be stored on the database
to give transparency and a visible trace on how was the game progress, and this object will be processed on our engine
in order to move forward on the game. When is a single player game, whenever a user creates a move, the AI will move 
immediately.

This is how a GameMove object looks on the database
```json
{
    "_id" : ObjectId("5c9d5c30e3872b0da34d3e7d"),
    "symbol" : "X",
    "player" : ObjectId("5c9d5c2ae3872b0da34d3e7a"),
    "cell" : {
        "row" : 0,
        "column" : 1
    },
    "created_at" : ISODate("2019-03-29T00:43:35.053Z"),
    "game" : ObjectId("5c9d5c2de3872b0da34d3e7c")
}
```
In order to trigger and manage moves we have the next endpoints:

#### POST /api/games/{game_id}
When posting to this url, while using the game_id which is the mongo generated id, it will then trigger a game
move based on the next json body:
```json
{
	"player": ObjectId("5c9d5c2ae3872b0da34d3e7a"),
	"symbol": "X",
	"cell": {
		"row": 0,
		"column": 1
	}
}
```
Where row and column are the coordinates of the cell on the grid where the marker is going to be placed. Almost everything
is validated, so the game can flow without issues. The symbol is the selected marker by the user. The same user cannot play
twice or the next user cannot use the same marker as the user before him.


#### GET /api/games/{game_id}/moves
This endpoint return the full list of moves for the given game_id.

## How to build
The project is using docker and docker-compose in order to be able to work on it locally.

It is enough with doing 
```
docker-compose up -d --build
```
And you will be able to access the application through `http://localhost:8000`

If you want to run the project locally with all its dependencies please follow the next steps:

* Install Pipenv if you don't have it already ([download here](https://pipenv.readthedocs.io/en/latest/)): $`pip install --user pipenv`
* Install mongodb locally or run it in a docker container.
* then run `pipenv install` on the project folder. This will create a virtual environment with all 
of the dependencies inside.
* Execute `cp .env.dist .env` to ensure the environment variables are loaded.
* To run the server just execute `pipenv run python main.py`. This will run the project on the port 8000.

Now in order to run the tests please run:
* Run `PIPENV_DOTENV_LOCATION=.env.test pipenv run python -m unittest app/tests/tests.py` in order to override
the .env file and use the testing configuration

## How to extend
The GameBot class uses an attribute called strategy. This strategy is the way the bot calculates its next move.
By default it uses the RandomStrategy built-in within the project, which just takes the available cells in the board
and picks a random one that it will be the bot next move. The goal is to even support implementations of machine learning 
algorithms (using reinforcement learning) into the calculation of the bot next move. The way to create new strategies is 
just to create a new class inheriting from app.engine.AbstractBotStrategy and then when instancing the bot, set the strategy
attribute to the new created strategy. The strategy must accept a board state, and the bot symbol to play by implementing the move method.

## Possible improvements
Due to time constraints there is a lot of room for improvement. One of the recognized improvements are:
* The way to ensure that the AI user is executed is on the Application creation. This should be part of the build process,
not of the application execution process.
* Better support for bots and association of a given bot with a given strategy.
* Proper separation between single player and multiplayer games.
* Support for Swagger and OpenAPI
* Better logging for debugging.

## Final words
The code is completely documented and the tests have a 90+% coverage. 

Im thankful for the opportunity and I hope this project gives you a better understanding of my 
skills and how do I like to design software.


CHANGELOG.md
# Changelog

## V1.1.0

- Style(engine): Pylint corrections. [Ricardo Santos Diaz]
- Fix(test): fix to unit test documentation on readme. [Ricardo Santos
  Diaz]
- Feat(game): Add support for multidimensional boards. [Ricardo Santos
  Diaz]
- Docs(project): Minor grammar fixes. [Ricardo Santos Diaz]
- Feat(bots): better support for strategy extension for next move
  calculation. [Ricardo Santos Diaz]
- Fix(games): Add back the retrieve for a single game method. [Ricardo
  Santos Diaz]
- Ci(prooject): Gitlab_ci implementation. [Ricardo Santos Diaz]
- Docs(project): Add the changelog. [Ricardo Santos Diaz]
- Build(project): Add docker-compose, change to env vars, documentation.
  [Ricardo Santos Diaz]
- Test(project): Add tests for engine, and api endpoints. [Ricardo
  Santos Diaz]
- Feat(project): Add the game engine and the ai. Add the endpoints.
  [Ricardo Santos Diaz]

  This is a huge update, first, all the endpoints are working as expected and specified. Regarding the game engine, or just the game rules, the engine module contains all of its elements.
- Feat(handlers): change handlers structure to be reusable, added
  decorators to validate. [Ricardo Santos Diaz]
- Feat(project): add handlers for users and games with get and post.
  [Ricardo Santos Diaz]
- Feat(project): add persistence layer with user and games object.
  [Ricardo Santos Diaz]

  There is currently an ODM helping with the persistence process. This ODM is umongo running on top of motor driver for mongo async operations. Based on these, two main classes were created: USer and Games. It was also added support to different configurations through environment variables.
- Build(project): Add the project initial dependencies and structure.
  [Ricardo Santos Diaz]


app/tests/tests.py
import asyncio
import json
import os
from typing import Tuple

from motor import MotorClient
from tornado.ioloop import IOLoop
from tornado.testing import AsyncHTTPTestCase
from app.urls import application
from app.models import User, Game
from main import ensure_ai_user


class BaseTest(AsyncHTTPTestCase):
    @classmethod
    def setUpClass(cls):
        cls.my_app = application

    def get_new_ioloop(self):
        return IOLoop.current()

    def get_app(self):
        return self.my_app


class TestRootApplication(BaseTest):

    def test_homepage(self,):
        response = self.fetch('/')
        self.assertEqual(response.code, 200)


class TestUserHandler(BaseTest):
    def get_app(self):
        db = MotorClient(
            os.getenv('DB_HOST', 'mongodb://localhost:27017'),
        )
        loop = asyncio.get_event_loop()
        loop.run_until_complete(db.drop_database(os.getenv('DB_NAME')))
        return application

    def create_user(self) -> str:
        user = User(username='usertest', email='test@test.de')
        loop = asyncio.get_event_loop()
        loop.run_until_complete(user.commit())
        return user.pk.__str__()

    def test_post_empty_json(self):
        response = self.fetch(
            '/api/users',
            method="POST",
            body="",
        )
        self.assertEqual(response.code, 400)

    def test_invalid_post(self):
        with open('app/tests/resources/user_invalid.json') as json:
            body = json.read()
        response = self.fetch(
            '/api/users',
            method="POST",
            body=body,
        )
        self.assertEqual(response.code, 500)

    def test_user_success(self):
        with open('app/tests/resources/user_success.json') as json:
            body = json.read()
        response = self.fetch(
            '/api/users',
            method="POST",
            body=body,
        )
        self.assertEqual(response.code, 200)

    def test_user_duplicated(self):
        self.create_user()
        with open('app/tests/resources/user_success.json') as json:
            body = json.read()
        response = self.fetch(
            '/api/users',
            method="POST",
            body=body,
        )
        self.assertEqual(response.code, 500)

    def test_get_user(self):
        user_id = self.create_user()
        response = self.fetch(
            '/api/users/%s' % user_id,
            method="GET",
        )
        self.assertEqual(response.code, 200)

    def test_get_user_invalid_mongo_id(self):
        response = self.fetch(
            '/api/users/%s' % '123456',
            method="GET",
        )
        self.assertEqual(response.code, 400)

    def test_get_user_not_found(self):
        mongo_id = '5c9d2b09e3872b287363cf28'
        response = self.fetch(
            '/api/users/%s' % mongo_id,
            method="GET",
        )
        self.assertEqual(response.code, 404)

    def test_update_user_not_found(self):
        with open('app/tests/resources/user_update.json') as json:
            body = json.read()
        user_id = '5c9d2b09e3872b287363cf28'
        response = self.fetch(
            '/api/users/%s' % user_id,
            method="PUT",
            body=body,
        )
        self.assertEqual(response.code, 404)

    def test_update_user_invalid_body(self):
        with open('app/tests/resources/user_update_invalid.json') as json:
            body = json.read()
        user_id = self.create_user()
        response = self.fetch(
            '/api/users/%s' % user_id,
            method="PUT",
            body=body,
        )
        self.assertEqual(response.code, 500)

    def test_update_user(self):
        with open('app/tests/resources/user_update.json') as json:
            body = json.read()
        user_id = self.create_user()
        response = self.fetch(
            '/api/users/%s' % user_id,
            method="PUT",
            body=body,
        )
        self.assertEqual(response.code, 200)

    def test_delete_user(self):
        user_id = self.create_user()
        response = self.fetch(
            '/api/users/%s' % user_id,
            method="DELETE",
        )
        self.assertEqual(response.code, 200)

    def test_delete_not_found(self):
        user_id = '5c9d2b09e3872b287363cf28'
        response = self.fetch(
            '/api/users/%s' % user_id,
            method="DELETE",
        )
        self.assertEqual(response.code, 404)

    def test_delete_success(self):
        user_id = self.create_user()
        response = self.fetch(
            '/api/users/%s' % user_id,
            method="DELETE",
        )
        self.assertEqual(response.code, 200)


class TestGameHandler(BaseTest):
    def get_app(self):
        db = MotorClient(
            os.getenv('DB_HOST', 'mongodb://localhost:27017'),
        )
        loop = asyncio.get_event_loop()
        loop.run_until_complete(db.drop_database(os.getenv('DB_NAME')))
        return application

    def create_player_one(self) -> str:
        user = User(username='playerone', email='player@one.te')
        loop = asyncio.get_event_loop()
        loop.run_until_complete(user.commit())
        return user.pk.__str__()

    def create_player_two(self) -> str:
        user = User(username='playertwo', email='player@two.te')
        loop = asyncio.get_event_loop()
        loop.run_until_complete(user.commit())
        return user.pk.__str__()

    def create_game(self) -> str:
        player_one_id = self.create_player_one()
        player_two_id = self.create_player_two()
        data = {"players": [player_one_id, player_two_id]}
        game = Game(**data)
        loop = asyncio.get_event_loop()
        loop.run_until_complete(game.commit())
        return game.pk.__str__()

    def test_post_empty_json(self):
        response = self.fetch(
            '/api/games',
            method="POST",
            body="",
        )
        self.assertEqual(response.code, 400)

    def test_invalid_post(self):
        with open('app/tests/resources/game_invalid.json') as json:
            body = json.read()
        response = self.fetch(
            '/api/games',
            method="POST",
            body=body,
        )
        self.assertEqual(response.code, 500)

    def test_game_success_multiplayer(self):
        player_one_id = self.create_player_one()
        player_two_id = self.create_player_two()
        data = {"players": [player_one_id, player_two_id]}
        response = self.fetch(
            '/api/games',
            method="POST",
            body=json.dumps(data, ensure_ascii=False),
        )
        self.assertEqual(response.code, 200)

    def test_game_success_single_player(self):
        player_one_id = self.create_player_one()
        data = {"players": [player_one_id]}
        response = self.fetch(
            '/api/games',
            method="POST",
            body=json.dumps(data, ensure_ascii=False),
        )
        self.assertEqual(response.code, 200)

    def test_get_game(self):
        game_id = self.create_game()
        response = self.fetch(
            '/api/games/%s' % game_id,
            method="GET",
        )
        self.assertEqual(response.code, 200)

    def test_get_game_invalid_mongo_id(self):
        response = self.fetch(
            '/api/games/%s' % '123456',
            method="GET",
        )
        self.assertEqual(response.code, 400)

    def test_get_game_not_found(self):
        game_id = '5c9d2b09e3872b287363cf28'
        response = self.fetch(
            '/api/games/%s' % game_id,
            method="GET",
        )
        self.assertEqual(response.code, 404)


class TestPlay(BaseTest):
    def get_app(self):
        db = MotorClient(
            os.getenv('DB_HOST', 'mongodb://localhost:27017'),
        )
        loop = asyncio.get_event_loop()
        loop.run_until_complete(db.drop_database(os.getenv('DB_NAME')))
        return application

    def create_player_one(self) -> str:
        user = User(username='playerone', email='player@one.te')
        loop = asyncio.get_event_loop()
        loop.run_until_complete(user.commit())
        return user.pk.__str__()

    def create_player_two(self) -> str:
        user = User(username='playertwo', email='player@two.te')
        loop = asyncio.get_event_loop()
        loop.run_until_complete(user.commit())
        return user.pk.__str__()

    def create_player_three(self) -> str:
        user = User(username='playerthree', email='player@three.te')
        loop = asyncio.get_event_loop()
        loop.run_until_complete(user.commit())
        return user.pk.__str__()

    def create_game(self) -> Tuple:
        player_one_id = self.create_player_one()
        player_two_id = self.create_player_two()
        data = {"players": [player_one_id, player_two_id]}
        game = Game(**data)
        loop = asyncio.get_event_loop()
        loop.run_until_complete(game.commit())
        return game.pk.__str__(), player_one_id, player_two_id

    def create_single_player_game(self) -> Tuple:
        ensure_ai_user()
        player_one_id = self.create_player_one()
        data = {"players": [player_one_id]}
        game = Game(**data)
        loop = asyncio.get_event_loop()
        loop.run_until_complete(game.commit())
        return game.pk.__str__(), player_one_id

    def test_game_move_invalid_move(self):
        game_id, player_one, player_two = self.create_game()
        data = {
            "player": player_one,
            "symbol": "X",
            "cell": {
                "row": 3,
                "column": 3
            }
        }
        response = self.fetch(
            '/api/games/%s' % game_id,
            method="POST",
            body=json.dumps(data, ensure_ascii=False),
        )
        self.assertEqual(response.code, 412)

    def test_valid_move(self):
        game_id, player_one, player_two = self.create_game()
        data = {
            "player": player_one,
            "symbol": "X",
            "cell": {
                "row": 0,
                "column": 0
            }
        }
        response = self.fetch(
            '/api/games/%s' % game_id,
            method="POST",
            body=json.dumps(data, ensure_ascii=False),
        )
        self.assertEqual(response.code, 200)

    def test_same_user_move(self):
        game_id, player_one, player_two = self.create_game()
        data = {
            "player": player_one,
            "symbol": "X",
            "cell": {
                "row": 0,
                "column": 0
            }
        }
        self.fetch(
            '/api/games/%s' % game_id,
            method="POST",
            body=json.dumps(data, ensure_ascii=False),
        )
        data = {
            "player": player_one,
            "symbol": "X",
            "cell": {
                "row": 0,
                "column": 0
            }
        }
        response = self.fetch(
            '/api/games/%s' % game_id,
            method="POST",
            body=json.dumps(data, ensure_ascii=False),
        )
        self.assertEqual(response.code, 412)

    def test_same_symbol(self):
        game_id, player_one, player_two = self.create_game()
        data = {
            "player": player_one,
            "symbol": "X",
            "cell": {
                "row": 0,
                "column": 0
            }
        }
        self.fetch(
            '/api/games/%s' % game_id,
            method="POST",
            body=json.dumps(data, ensure_ascii=False),
        )
        data = {
            "player": player_two,
            "symbol": "X",
            "cell": {
                "row": 0,
                "column": 0
            }
        }
        response = self.fetch(
            '/api/games/%s' % game_id,
            method="POST",
            body=json.dumps(data, ensure_ascii=False),
        )
        self.assertEqual(response.code, 412)

    def test_same_cell(self):
        game_id, player_one, player_two = self.create_game()
        data = {
            "player": player_one,
            "symbol": "X",
            "cell": {
                "row": 0,
                "column": 0
            }
        }
        self.fetch(
            '/api/games/%s' % game_id,
            method="POST",
            body=json.dumps(data, ensure_ascii=False),
        )
        data = {
            "player": player_two,
            "symbol": "O",
            "cell": {
                "row": 0,
                "column": 0
            }
        }
        response = self.fetch(
            '/api/games/%s' % game_id,
            method="POST",
            body=json.dumps(data, ensure_ascii=False),
        )
        self.assertEqual(response.code, 412)

    def test_second_move(self):
        game_id, player_one, player_two = self.create_game()
        data = {
            "player": player_one,
            "symbol": "X",
            "cell": {
                "row": 0,
                "column": 0
            }
        }
        self.fetch(
            '/api/games/%s' % game_id,
            method="POST",
            body=json.dumps(data, ensure_ascii=False),
        )
        data = {
            "player": player_two,
            "symbol": "O",
            "cell": {
                "row": 1,
                "column": 1
            }
        }
        response = self.fetch(
            '/api/games/%s' % game_id,
            method="POST",
            body=json.dumps(data, ensure_ascii=False),
        )
        self.assertEqual(response.code, 200)

    def test_single_player(self):
        game_id, player_one = self.create_single_player_game()
        data = {
            "player": player_one,
            "symbol": "X",
            "cell": {
                "row": 0,
                "column": 0
            }
        }
        response = self.fetch(
            '/api/games/%s' % game_id,
            method="POST",
            body=json.dumps(data, ensure_ascii=False),
        )
        print(response.body.decode())
        self.assertEqual(response.code, 200)


app/urls.py
"""
Collection of project routes and where do they point to
"""
import os
from tornado.web import Application, url
from app.models import User, Game, GameMove
import app.handlers


application = Application([
        url(
            r"/",
            app.handlers.MainHandler,
        ),
        url(
            r"/api/users",
            app.handlers.AbstractGeneralHandler,
            {'cls': User},
        ),
        url(
            r"/api/users/(?P<pk>\w+)",
            app.handlers.AbstractObjHandler,
            {'cls': User},
        ),
        url(
            r"/api/games",
            app.handlers.AbstractGeneralHandler,
            {'cls': Game}
        ),
        url(
            r"/api/games/(?P<pk>\w+)",
            app.handlers.GameMoveHandler,
        ),
        url(
            r"/api/games/(?P<pk>\w+)/moves",
            app.handlers.GameMovesRetriever,
        ),
    ],
    debug=os.getenv('DEBUG_FLAG') == "True",
)


app/handlers.py
"""
Collection of project handlers
"""
import json
from tornado.web import RequestHandler, HTTPError
from tornado.escape import json_decode
from umongo import fields

from app.decorators import validate_mongo_id, validate_json_body
from app.models import Game, GameMove
from app.engine import GameEngine, GameBot


engine = GameEngine()

class MainHandler(RequestHandler):
    """
    Main handler to return the default responses
    """
    async def get(self):
        """
        Base handler, hello world
        :return:
        """
        self.set_header("Content-Type", 'application/json')
        self.write({"Hello": "World"})


class ErrorHandler(RequestHandler):
    """
    This class allows us to handle the HTTPErrors and exceptions
    """
    def write_error(self, status_code, **kwargs):
        """
        This method takes the HTTPError and renders a json
        based on it
        :param status_code:
        :param kwargs:
        :return:
        """
        self.set_header('Content-Type', 'application/json')
        status_code = 500
        if "exc_info" in kwargs:
            exception = kwargs.get('exc_info')[1]
            self._reason = exception.__str__()
            status_code = getattr(exception, 'status_code', 400)
        self.finish(json.dumps({
            'error': {
                'code': status_code,
                'message': self._reason,
            }
        }))


class AbstractGeneralHandler(ErrorHandler):
    """
    General operations that require no objectId,
    This includes POST and GET list operations.
    """
    def initialize(self, cls):
        """
        Initialize the class variable with ODM definition
        :param cls: umongo ODM class definition
        :return:
        """
        self.cls = cls

    @validate_json_body
    async def post(self):
        """
        Create a new object by the POST json body
        :return:
        """
        data = json_decode(self.request.body)
        obj = self.cls(**data)
        await obj.ensure_indexes()
        await obj.commit()
        self.set_header("Content-Type", 'application/json')
        self.write(obj.dump())

    async def get(self):
        """
        Retrieve a list of objects
        :return:
        """
        cursor = self.cls.find()
        data = []
        async for obj in cursor:
            data.append(obj.dump())
        self.set_header("Content-Type", 'application/json')
        self.write({'data': data})


class AbstractObjHandler(ErrorHandler):
    """
    This handler has the methods to manipulate a single object
    based on its ObjectId
    """
    def initialize(self, cls):
        """
        Initialize the class variable with ODM definition
        :param cls: umongo ODM class definition
        :return:
        """
        self.cls = cls

    @validate_mongo_id
    async def get(self, pk: str):
        """
        Retrieve a single object from the database
        :param str pk:
        :return:
        """
        obj = await self.cls.find_by_id(pk)
        self.set_header("Content-Type", 'application/json')
        self.write(obj.dump())

    @validate_mongo_id
    @validate_json_body
    async def put(self, pk: str):
        """
        Edit a given database object
        :param str pk:
        :return:
        """
        obj = await self.cls.find_by_id(pk)
        data = json_decode(self.request.body)
        obj.update(data)
        await obj.commit()
        self.set_header("Content-Type", 'application/json')
        self.write(obj.dump())

    @validate_mongo_id
    async def delete(self, pk: str):
        """
        Remove a given database object
        :param str pk:
        :return:
        """
        obj = await self.cls.find_by_id(pk)
        await obj.delete()
        self.set_header("Content-Type", 'application/json')
        self.write({'success': True})


class GameMoveHandler(ErrorHandler):
    """
    This class allows us to trigger execute board move
    operations
    """

    @validate_mongo_id
    @validate_json_body
    async def post(self, pk: str):
        """
        Trigger a game move by the user. If the game is multiplayer
        will execute a bot move immediatly after. This can be vastly
        improved by the use of the strategy pattern to distinguish
        single player and multiplayer games. Right now in order to not
        increase complexity, and as the game rules are no different on both
        modes we are going by a simple decision.
        :param str pk:
        :return:
        """
        game: Game = await Game.find_by_id(pk)
        if game.status in [Game.STATUS_TIE, Game.STATUS_FINISHED]:
            raise HTTPError(
                400,
                'Game was already finished'
            )
        data = json_decode(self.request.body)
        move = GameMove(**data)
        await engine.execute_move(game, move)
        if not game.multiplayer and game.status == Game.STATUS_IN_PROGRESS:
            aimove = await GameBot().move(game, move)
            await engine.execute_move(game, aimove)
        self.set_header("Content-Type", 'application/json')
        self.write(game.dump())

    @validate_mongo_id
    async def get(self, pk: str):
        """
        Retrieve a single object from the database
        :param str pk:
        :return:
        """
        obj = await Game.find_by_id(pk)
        self.set_header("Content-Type", 'application/json')
        self.write(obj.dump())


class GameMovesRetriever(ErrorHandler):
    """
    This handler allows us to retrieve the moves trace for a
    given game.
    """
    @validate_mongo_id
    async def get(self, pk: str):
        """
        Retrieve a list of moves from the database
        :param str pk:
        :return:
        """

        cursor = GameMove.find({"game": fields.ObjectId(pk)})
        data = []
        async for obj in cursor:
            data.append(obj.dump())
        self.set_header("Content-Type", 'application/json')
        self.write({"data": data})


app/decorators.py
"""
Set of decorators useful for the main tool handlers
"""
import bson
from tornado.web import HTTPError


def validate_mongo_id(func):
    """
    Validate that the input contains a valid mongoId
    :param func:
    :return:
    """
    def wrapper(*args, **kwargs):
        if not bson.objectid.ObjectId.is_valid(kwargs.get('pk', None)):
            raise HTTPError(400, 'Invalid Mongo Id')
        return func(*args, **kwargs)
    return wrapper


def validate_json_body(func):
    """
    Validate that the body of the request is not empty
    If is empty throw a 400 exception
    :param func:
    :return:
    """
    def wrapper(*args, **kwargs):
        if not args[0].request.body:
            raise HTTPError(400, 'Empty request body')
        return func(*args, **kwargs)
    return wrapper


app/models.py
"""
Module with the object definitions that are going to be stored
"""
import os
import datetime
from motor.motor_asyncio import AsyncIOMotorClient
from tornado.web import HTTPError
from umongo import Instance, \
    Document, \
    fields, \
    validate, \
    ValidationError


db = AsyncIOMotorClient(
    os.getenv('DB_HOST', 'mongodb://localhost:27017'),
)[os.getenv('DB_NAME', 'tictactoe')]
instance = Instance(db)


@instance.register
class BaseDocument(Document):
    """
    This class looks to implement a common method on all of our
    produced documents.
    """

    @classmethod
    async def find_by_id(cls, pk: str):
        """
        Abstraction to throw an exception in case an object
        does not exist on db.
        :param str pk:
        :return:
        """
        obj = await cls.find_one(
            {"_id": fields.ObjectId(pk)}
        )
        if obj is None:
            raise HTTPError(
                404,
                'Object not found Not Found',
            )
        return obj

    class Meta:
        """
        ODM Metadata
        """
        allow_inheritance = True
        abstract = True


@instance.register
class User(BaseDocument):
    """
    This document stores the user information and data
    as the number of victories and defeats
    """
    username = fields.StrField(
        required=True,
        unique=True,
        validate=[validate.Length(min=6, max=20)],
    )

    email = fields.StrField(
        required=True,
        unique=True,
        validate=validate.Email(),
    )

    victories = fields.IntField(
        default=0,
    )

    created_at = fields.DateTimeField(
        default=datetime.datetime.now(),
    )

    class Meta:
        """
        ODM Metadata
        """
        collection = db.users


@instance.register
class Game(BaseDocument):
    """
    This is how we are going to store the game
    """
    STATUS_CREATED = 'created'
    STATUS_IN_PROGRESS = 'in_progress'
    STATUS_TIE = 'tie'
    STATUS_FINISHED = 'finished'

    STATUS = [
        STATUS_CREATED,
        STATUS_IN_PROGRESS,
        STATUS_TIE,
        STATUS_FINISHED,
    ]

    players = fields.ListField(
        fields.ReferenceField("User"),
    )

    multiplayer = fields.BoolField(
        default=False,
    )

    board = fields.ListField(
        fields.ListField(
            fields.StrField()
        )
    )

    status = fields.StrField(
        validate=validate.OneOf(STATUS),
        default=STATUS_CREATED
    )

    created_at = fields.DateTimeField(
        default=datetime.datetime.now(),
    )

    winner = fields.ReferenceField("User")

    size = fields.IntegerField(
        validate=[validate.Range(3, 10)],
        default=3,
    )

    def pre_insert(self):
        """
        Fill the board and do multiplayer validations
        """
        if not self.players:
            raise ValidationError(
                "There should be at least one player",
            )
        if len(self.players) > 2:
            raise ValidationError(
                "Maximum two players"
            )
        if len(self.players) == 2:
            self.multiplayer = True
        else:
            self.multiplayer = False

        self.board = [["" for i in range(self.size)] for y in range(self.size)]
        pass

    class Meta:
        """
        ODM Metadata
        """
        collection = db.games


@instance.register
class GameMove(BaseDocument):
    """
    This collection will track the moves of the users
    on the given game
    """

    SYMBOLS = [
        'X',
        'O'
    ]

    game = fields.ReferenceField(
        "Game",
    )

    cell = fields.DictField(
        required=True
    )

    symbol = fields.StrField(
        validate=validate.OneOf(SYMBOLS),
        required=True,
    )

    player = fields.ReferenceField(
        "User",
        required=True,
    )

    created_at = fields.DateTimeField(
        default=datetime.datetime.now(),
    )

    class Meta:
        """
        ODM Metadata
        """
        collection = db.moves
